packages/core/src/common/message-rpc/uint8-array-message-buffer.ts

```diff
@@ -0,0 +1,206 @@
+// *****************************************************************************
+// Copyright (C) 2021 Red Hat, Inc. and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// http://www.eclipse.org/legal/epl-2.0.
+//
+// This Source Code may also be made available under the following Secondary
+// Licenses when the conditions for such availability set forth in the Eclipse
+// Public License v. 2.0 are satisfied: GNU General Public License, version 2
+// with the GNU Classpath Exception which is available at
+// https://www.gnu.org/software/classpath/license.html.
+//
+// SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+// *****************************************************************************
+import { Disposable } from 'vscode-languageserver-protocol';
+import { Emitter, Event } from '../event';
+import { ReadBuffer, WriteBuffer } from './message-buffer';
+
+/**
+ * The default {@link WriteBuffer} implementation. Uses a {@link Uint8Array} for buffering.
+ * The {@link Uint8ArrayWriteBuffer.onCommit} hook can be used to rect to on-commit events.
+ * After the {@link Uint8ArrayWriteBuffer.commit} method has been called the buffer is disposed
+ * and can no longer be used for writing data. If the writer buffer is no longer needed but the message
+ * has not been committed yet it has to be disposed manually.
+ */
+export class Uint8ArrayWriteBuffer implements WriteBuffer, Disposable {
+
+    private encoder = new TextEncoder();
+    private msg: DataView;
+    private isDisposed = false;
+    private offset: number;
+
+    constructor(private buffer: Uint8Array = new Uint8Array(1024 * 1024), writePosition: number = 0) {
```

Comment:
I'm wondering whether whether the default should be smaller. How many of these write buffers do we allocate?

Reply:
Hard to say, depends on the amount of concurrently processed rpc-messages. Typically these buffers are short-lived and should be garbage collected once the message has been comitted (or the buffer was disposed manually).
Nevertheless, I did some testing and reducing the default value (to e.g. 1024) has no significant impact on encoding times . So using a smaller default value is definitely a good idea to avoid unnecessary memory allocation. I have pushed an update that sets the default size to "1024".